### ISSUE TITLE
SM89 (Ada): concurrent long-context prefill (varlen batch) corrupts output - undocumented limitation

### ISSUE THREAD

**Issue by yzxiu** [yhfgyyf/vllm-deepseek-v4-sm89#57]
## Summary

**SM89 (Ada): concurrent long-context prefill — varlen batch corrupts output (wrong numbers / prompt repetition / refusals) when `long_prefill_token_threshold` allows >1 request to interleave in one scheduler step.**

The README validates everything under **single concurrency**, so this path is never exercised and the limitation is undocumented. I hit it in production (multi-user long-context serving) and want to flag it + get confirmation it's a known/intended constraint on SM89.

## Environment

- HW: 4× RTX 5880 Ada (SM89, cap 8.9), 48 GB each, CUDA 13.0 toolchain, torch 2.11.0+cu130
- vLLM fork wheel + FlashInfer 0.6.14 sparse MLA JIT (as per install instructions)
- Model: `DeepSeek-V4-Flash-0731`
- k8s deployment, TP=4, `--kv-cache-dtype fp8_ds_mla`, `--block-size 256`, `--no-enable-prefix-caching`, `--max-num-seqs 16`

## Reproduction

Request = ~50K-token context (retrieval: "What is SECRET_NUMBER in paragraph N?"), 4 concurrent workers, 12 requests, `temperature=0`, max_tokens=16.

Vary these two scheduler knobs:

| `max-num-batched-tokens` | `long-prefill-token-threshold` | interleaving | accuracy |
|---|---|---|---|
| 2048 (default) | default (auto = ~4% of max_len) | none (1 req/step) | **100%** |
| 4096 | 4096 | none (1 req/step, chunk 4K) | **100%** |
| 4096 | 2048 | 2 reqs interleave | **41.7%** (5/12) |
| 2048 | 512 | 4 reqs interleave | **8.3%** (1/12) |

The moment 2+ long requests interleave their chunks in one scheduler step (i.e. a **varlen batch**), accuracy collapses. Failures are: prompt repetition ("Paragraph 751: ..."), refusals ("I cannot fulfill this request"), and wrong numbers. No exceptions in the engine log — silently wrong output.

Engine log during interleaved runs shows `Running: 4 reqs` (parallel prefill confirmed), and `DSA indexer decode path: use_flattening=True supports_varlen=False (next_n=1, ...)`.

## Root cause (my reading)

`vllm/v1/attention/backends/mla/indexer.py`:

```python
def _supports_varlen_paged_mqa_logits() -> bool:
    return (
        current_platform.is_cuda()
        and current_platform.is_device_capability_family(100)   # SM100 only
        and has_deep_gemm()
    )
```

- `get_cudagraph_support` → `UNIFORM_BATCH` on SM89 (vs `ALWAYS` on SM100): CUDA-graph capture only supports equal-length batches.
- `supports_varlen = _supports_varlen_paged_mqa_logits()` → `False` on SM89.

So concurrent long prefill == varlen batch == the unsupported path, and the SM89 fallback silently produces wrong sparse MLA indexing — corrupting output with no error. Single-request serving (equal-length, uniform batch) is always correct, which is why the README's single-concurrency tests all pass.

## Ask

1. Is this a **known/intended constraint** on SM89? If so, could the README document it (the README currently states SM89 prefill/decode is "adapted" and validated, without noting this concurrency limitation)?
2. Is there a supported way to get **correct concurrent long-context prefill on SM89** — e.g. padding requests to uniform batch so the UNIFORM_BATCH cudagraph path applies, or a fallback dense indexer for varlen batches?
3. If none, confirm that on SM89 correct long-pre fill is inherently single-stream (1 request per scheduler step) and concurrent long context is out of scope.

Happy to provide the full test scripts and logs if useful.

**Comment by wahhhhhh999** [comment]
请问该问题解决了吗？目前我们在l40上部署的也出现了回答重复和乱码的情况

**Comment by yzxiu** [comment]
> 请问该问题解决了吗？目前我们在l40上部署的也出现了回答重复和乱码的情况

很有可能是并发导致的问题，加上 --no-enable-prefix-caching 试试
